TESTS/mbed_hal/watchdog_timing/main.cpp

```diff
@@ -0,0 +1,150 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2018 ARM Limited
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#if !DEVICE_WATCHDOG
+#error [NOT_SUPPORTED] Watchdog not supported for this target
+#endif
+
+#include "greentea-client/test_env.h"
+#include "hal/watchdog_api.h"
+#include "unity/unity.h"
+#include "us_ticker_api.h"
+#include "utest/utest.h"
+#include "watchdog_timing_tests.h"
+
+#define MSG_VALUE_DUMMY "0"
+#define CASE_DATA_INVALID 0xffffffffUL
+
+#define MSG_VALUE_LEN 24
+#define MSG_KEY_LEN 24
+
+#define MSG_KEY_DEVICE_READY "ready"
+#define MSG_KEY_START_CASE "start_case"
+#define MSG_KEY_HEARTBEAT "hb"
+
+using utest::v1::Case;
+using utest::v1::Specification;
+using utest::v1::Harness;
+
+struct testcase_data {
+    int index;
+    int start_index;
+    uint32_t received_data;
+};
+
+testcase_data current_case;
+
+template<uint32_t timeout_ms, uint32_t delta_ms>
+void test_timing()
+{
+    // Phase 2. -- verify the test results.
+    // Verify the heartbeat time span sent by host is within given delta.
+    if (current_case.received_data != CASE_DATA_INVALID) {
+        TEST_ASSERT_UINT32_WITHIN(delta_ms, timeout_ms, current_case.received_data);
+        current_case.received_data = CASE_DATA_INVALID;
+        return;
+    }
+
+    // Phase 1. -- run the test code.
```

Comment:
Is it normal that phase 1 is after phase 2 ?

Reply:
I think these get called twice - so phase 1 always happens first (before reset) then phase 2 (post-reset)

Reply:
Correct. The phase 2 is executed when the host sends a valid time measurement after the device reset.